### ISSUE TITLE
Include replica of iMessage curved lines connecting replies 

### ISSUE THREAD

**Issue by ElrB** [ekazaev/ChatLayout#66]
Hi @ekazaev 

I have tried to implement the solution you pointed out in the last issue, but without any good outcome. 
Below I will add screenshot of how I have went on to implement the solution and results.

New variables in MainContainerView class:
![Screenshot 2024-02-08 at 7 18 07 PM](https://github.com/ekazaev/ChatLayout/assets/66596779/73eea812-d9c2-49d6-85ba-b7a5e4ea40d2)

- setupSubviews()
![Screenshot 2024-02-08 at 7 20 50 PM](https://github.com/ekazaev/ChatLayout/assets/66596779/e017997a-fcf5-412c-8f96-12767d944779)

- updateContainerView() & resetContainerView()
![Screenshot 2024-02-08 at 7 22 08 PM](https://github.com/ekazaev/ChatLayout/assets/66596779/c4140f28-af13-40b5-a181-b96e4d5282b4)

- CurvedLineView
![Screenshot 2024-02-08 at 7 23 03 PM](https://github.com/ekazaev/ChatLayout/assets/66596779/6a1885e9-2662-4f88-a37f-f29ea347cc65)

- Data source createTextCell (here I check if message is a reply so that curve line view can be created or not)
![Screenshot 2024-02-08 at 7 23 51 PM](https://github.com/ekazaev/ChatLayout/assets/66596779/5212ef81-f661-435f-9c62-22eccf019ed2)

Results:

It seems like the width of the line is just off, even if I try get the width of the bubble view then subtract it from the screen size so that I get get the remaining width that the bubble view is not occupying. When I scroll, weird behaviours start occurring with the curve line, it starts changing in width and height size, affect the bubble view size as well. The weird behaviours make me think if reuse of cells without clearing the curved line is the cause, but not sure.

Unexpected result marked with red, expected result marked with blue. 

![image_6487327-3](https://github.com/ekazaev/ChatLayout/assets/66596779/cc4a0f5c-0bdf-4c2e-b63a-2e46fa327d83)
![image_6487327-4](https://github.com/ekazaev/ChatLayout/assets/66596779/86cb8ac7-4958-456c-9e86-27e672465238)
![image_6487327-5](https://github.com/ekazaev/ChatLayout/assets/66596779/589b45fc-a51e-4977-8ef9-4abdded7a26e)

Please help. Thanks.

**Comment by ekazaev** [comment]
@ElrB Unfortunately I cant write code for you. I can give you only some directions. It looks like you are on the right path but I cant sit and debug your code. I am sorry. 

If I can suggest something: take constraints from the layout like TextMessageView does
```
    func apply(_ layoutAttributes: ChatLayoutAttributes) {
        viewPortWidth = layoutAttributes.layoutFrame.width
        setupSize()
    }
```
Pass it somehow (create a multiple delegate listener for example) and pass it up to the all views that might be interested. In your example It is unclear why do you set up width to `310`. Iphones have different screen sizes.

Here it seems that the cell is not handling reuse properly. Because there is no reply in the previous cell.
https://private-user-images.githubusercontent.com/66596779/303429475-589b45fc-a51e-4977-8ef9-4abdded7a26e.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc0MjQ3NzYsIm5iZiI6MTcwNzQyNDQ3NiwicGF0aCI6Ii82NjU5Njc3OS8zMDM0Mjk0NzUtNTg5YjQ1ZmMtYTUxZS00OTc3LThlZjktNGFiZGRlZDdhMjZlLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA4VDIwMzQzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmYTVkMDI1NDY2MmRiZmIyNjYxZGZhOWVjMGQ2ZGJkMGZkMDQyZWMwZmE2YmNiNTk3ODc3NGY3YWUwMTE0MzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.l8RSAVO-inFM2c1qkJrPBtzVNPR-OAm6QsiD7K2diyU

I see you decided not to move the reply bubble into the message bubble. That means that you also need to pass somehow the information where is the reply in the previous cell is positioned and act accordingly. I am also not sure how you are going to handle the situation is the reply bubble is short. Even if the curve will work correctly it will point into the void, not the best visual decision imho.

That is as much as I can help you unfortunately. It hard to look at the code without an ability to put a breakpoint and check the constraints and values.

Keep trying. 

Best of luck.

**Comment by ElrB** [comment]
Hi @ekazaev 

I will leave this route, and take the WhatsApp reply style.
I just have one question or seek your last help then we can close the issue.

Right now I see bubble view is create like; let bubbleView = cell.customView.customView.customView in the createCell methods.

How can create a BezierMaskedView<CustomView> (message bubble view) without having to dequeue/initialise a cell?

Thanks.

**Comment by ekazaev** [comment]
@ElrB 

> How can I create a BezierMaskedView (message bubble view) without having to dequeue/initialise a cell?

Do you mean like this?

```swift
let view =  BezierMaskedView<UIView>(frame: CGRect(origin: .zero, size: .init(width: 300, height: 50)))
view.customView.backgroundColor = .green
```
<img width="339" alt="Screenshot 2024-02-12 at 14 26 11" src="https://github.com/ekazaev/ChatLayout/assets/20243258/53a5684a-b003-4b4a-a6e4-0c751350195e">

The second I do not understand. I need to see the code to try. Fork the example app, modify it and Ill have a look.

But I suggest you to use an extra container view rather than modifying the `TextMessageView`. I a similar way like I added an extra stack view in the container in your previous question.

**Comment by ElrB** [comment]
Hi @ekazaev 

Thanks for answering the first part. Feel a bit dumb now :D

For the second part. Try add a constant for the text view top anchor like in below image, then scroll the chat up and down. The constraints are not consistent after scrolling. I want to push the text view down so that I can make space to add the reply message view. I tried activating the top anchor constraint in reloadData method but issue was still there.

![Screenshot 2024-02-12 at 6 30 03 PM](https://github.com/ekazaev/ChatLayout/assets/66596779/39072764-85ba-4570-947f-3cd00762b843)

**Comment by ekazaev** [comment]
@ElrB I do not see such behaviour. Everything is as expected.

<img width="565" alt="Screenshot 2024-02-12 at 17 06 42" src="https://github.com/ekazaev/ChatLayout/assets/20243258/3ae5d441-df0c-47fd-916e-b62a72343524">

```
diff --git a/Example/ChatLayout/Chat/View/Text Message View/TextMessageView.swift b/Example/ChatLayout/Chat/View/Text Message View/TextMessageView.swift
index c9a704f..6d226fb 100644
--- a/Example/ChatLayout/Chat/View/Text Message View/TextMessageView.swift	
+++ b/Example/ChatLayout/Chat/View/Text Message View/TextMessageView.swift	
@@ -23,6 +23,7 @@ final class TextMessageView: UIView, ContainerCollectionViewCellDelegate {
     private var controller: TextMessageController?
 
     private var textViewWidthConstraint: NSLayoutConstraint?
+    private var textViewTopConstraint: NSLayoutConstraint?
 
     override init(frame: CGRect) {
         super.init(frame: frame)
@@ -74,7 +75,8 @@ final class TextMessageView: UIView, ContainerCollectionViewCellDelegate {
         guard let controller else {
             return
         }
-        textView.text = controller.text
+        textView.text = "\(controller.text.count % 2 == 0 ? "[Odd]" : "[Even]") \(controller.text)"
+        setupSize()
         UIView.performWithoutAnimation {
             if #available(iOS 13.0, *) {
                 textView.textColor = controller.type.isIncoming ? UIColor.label : .systemBackground
@@ -112,18 +114,25 @@ final class TextMessageView: UIView, ContainerCollectionViewCellDelegate {
         textView.font = UIFont.preferredFont(forTextStyle: .body)
         addSubview(textView)
         NSLayoutConstraint.activate([
-            textView.topAnchor.constraint(equalTo: layoutMarginsGuide.topAnchor),
             textView.bottomAnchor.constraint(equalTo: layoutMarginsGuide.bottomAnchor),
             textView.leadingAnchor.constraint(equalTo: layoutMarginsGuide.leadingAnchor),
             textView.trailingAnchor.constraint(equalTo: layoutMarginsGuide.trailingAnchor)
         ])
+        textViewTopConstraint = textView.topAnchor.constraint(equalTo: layoutMarginsGuide.topAnchor, constant: 80)
+        textViewTopConstraint?.priority = UILayoutPriority(rawValue: 999)
         textViewWidthConstraint = textView.widthAnchor.constraint(lessThanOrEqualToConstant: viewPortWidth)
         textViewWidthConstraint?.isActive = true
+        textViewTopConstraint?.isActive = true
     }
 
     private func setupSize() {
         UIView.performWithoutAnimation {
             self.textViewWidthConstraint?.constant = viewPortWidth * Constants.maxWidth
+            if (controller?.text.count ?? 0) % 2 == 0 {
+                textViewTopConstraint?.constant = 80
+            } else {
+                textViewTopConstraint?.constant = 0
+            }
             setNeedsLayout()
         }
     }
```

**Comment by ElrB** [comment]
Thanks @ekazaev 
I think am good to finish coding replies now.
